tests/components/sonos/test_media_player.py

```diff
@@ -272,3 +278,109 @@ async def test_play_media_music_library_playlist_dne(
     assert soco_mock.play_uri.call_count == 0
     assert media_content_id in caplog.text
     assert "playlist" in caplog.text
+
+
+@pytest.mark.parametrize(
+    ("source", "result"),
+    [
+        (
+            SOURCE_LINEIN,
+            {
+                "switch_to_line_in": 1,
+            },
+        ),
+        (
+            SOURCE_TV,
+            {
+                "switch_to_tv": 1,
+            },
+        ),
+        (
+            "James Taylor Radio",
+            {
+                "play_uri": 1,
+                "play_uri_uri": "x-sonosapi-radio:ST%3aetc",
+                "play_uri_title": "James Taylor Radio",
+            },
+        ),
+        (
+            "66 - Watercolors",
+            {
+                "play_uri": 1,
+                "play_uri_uri": "x-sonosapi-hls:Api%3atune%3aliveAudio%3ajazzcafe%3aetc",
+                "play_uri_title": "66 - Watercolors",
+            },
+        ),
+        (
+            "1984",
+            {
+                "add_to_queue": 1,
+                "add_to_queue_item_id": "A:ALBUMARTIST/Aerosmith/1984",
+                "clear_queue": 1,
+                "play_from_queue": 1,
+            },
+        ),
+    ],
+)
+async def test_select_source(
+    hass: HomeAssistant,
+    soco_factory: SoCoMockFactory,
+    async_autosetup_sonos,
+    source: str,
+    result: dict[str, Any],
+) -> None:
+    """Test the select_source method with a variety of inputs."""
+    soco_mock = soco_factory.mock_list.get("192.168.42.2")
+
+    await hass.services.async_call(
+        MP_DOMAIN,
+        SERVICE_SELECT_SOURCE,
+        {
+            "entity_id": "media_player.zone_a",
+            "source": source,
+        },
+        blocking=True,
+    )
+
+    assert soco_mock.switch_to_line_in.call_count == result.get("switch_to_line_in", 0)
+    assert soco_mock.switch_to_tv.call_count == result.get("switch_to_tv", 0)
+    assert soco_mock.clear_queue.call_count == result.get("clear_queue", 0)
+    if result.get("add_to_queue"):
```

Comment:
Please remove the if statements from the test. Only one test needs "add_to_queue" and "play_from_queue".
Also you can create a common function to use in multiple tests

Reply:
This was good advice, and apologies for not getting all of them out after your last review. In the process of simplifying I found one missing mock call that should have been tested but wasn't - and it was obscured due to trying to be overly clever in the tests.